src/dev/run_check_core_api_changes.ts

```diff
@@ -0,0 +1,187 @@
+/*
+ * Licensed to Elasticsearch B.V. under one or more contributor
+ * license agreements. See the NOTICE file distributed with
+ * this work for additional information regarding copyright
+ * ownership. Elasticsearch B.V. licenses this file to you under
+ * the Apache License, Version 2.0 (the "License"); you may
+ * not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *    http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing,
+ * software distributed under the License is distributed on an
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ * KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+
+import { Extractor, IExtractorConfig } from '@microsoft/api-extractor';
+import chalk from 'chalk';
+import dedent from 'dedent';
+import execa from 'execa';
+import getopts from 'getopts';
+import { run } from './run';
+
+const apiExtractorConfig: IExtractorConfig = {
+  compiler: {
+    configType: 'tsconfig',
+    rootFolder: '.',
+  },
+  project: {
+    entryPointSourceFile: 'target/types/type_exports.d.ts',
+  },
+  apiReviewFile: {
+    enabled: true,
+    apiReviewFolder: './common/core_api_review/',
+    tempFolder: './build',
+  },
+  apiJsonFile: {
+    enabled: true,
+    outputFolder: './build',
+  },
+  dtsRollup: {
+    enabled: false,
+    trimming: true,
+    publishFolderForInternal: '',
+    publishFolderForBeta: '',
+    publishFolderForPublic: '',
+    mainDtsRollupPath: '',
+  },
+};
+
+const runBuildTypes = async () => {
+  await execa.shell('yarn run build:types');
+};
+
+const runApiDocumenter = async () => {
+  await execa.shell('api-documenter markdown -i ./build -o ./docs/development/core/api');
+};
+
+const runApiExtractor = (acceptChanges: boolean = false, log: any) => {
+  // Because of the internals of api-extractor ILogger can't be implemented as a typescript Class
+  const warnings = [] as string[];
+  const errors = [] as string[];
+
+  const memoryLogger = {
+    logVerbose(message: string) {
+      return null;
+    },
+
+    logInfo(message: string) {
+      return null;
+    },
+
+    logWarning(message: string) {
+      warnings.push(message);
+    },
+
+    logError(message: string) {
+      errors.push(message);
+    },
+  };
+
+  const options = {
+    // Indicates that API Extractor is running as part of a local build,
+    // e.g. on developer's machine. For example, if the *.api.ts output file
+    // has differences, it will be automatically overwritten for a
+    // local build, whereas this should report an error for a production build.
+    localBuild: acceptChanges,
+    customLogger: memoryLogger,
+  };
+
+  const extractor = new Extractor(apiExtractorConfig, options);
+  extractor.processProject();
+
+  errors.forEach(msg => log.error(msg));
```

Comment:
according to https://github.com/Microsoft/web-build-tools/blob/master/apps/api-extractor/src/api/ILogger.ts any error or warning will break prod build. should we do `process.exit(1)` on any such message, not just `apiChanged`?

Reply:
I've made it more permissible than their default so that warnings about missing documentation tags or incorrect TSDoc comments doesn't break the build. 

But perhaps it's better to start with a strict implementation and only relax it if we realise it is introducing annoying problems.